### PR TITLE
Helper for delay in milliseconds

### PR DESCRIPTION
**Pull request by prashanth-92**
#232 Attempts to provide a helper for adding a delay in milliseconds. The test shows a way to use the method.

### CODE REVIEW

**Comment by ctimmerm** [comment]
Hi @prashanth-92! Thanks for your PR, but unfortunately your change doesn't actually introduce per-request delays. The added method instead sets the delay for all requests, so for example:
```js
mock.delayInMs(200).onGet("/foo").reply(200);
mock.delayInMs(0).onGet("/bar").reply(200);
```
will cause no delay when a request to `/foo` is made.

**Comment by prashanth-92** [comment]
Hey @ctimmerm, yes my solution is setting the delay in the `MockAdapter` directly. Just to see if I get this right, if we need to add it at an endpoint level, then we might need to introduce a state in the handlers per verb and then use delay it with a promise approach you mentioned in one of the comments.

**Comment by prashanth-92** [comment]
@ctimmerm I have a revised approach with delay per request. Can you please review it?

**Comment by prashanth-92** [comment]
Ping @ctimmerm If you have some time, please review this pull request

File: src/handle_request.js [hunk 1]
```diff
@@ -137,4 +137,12 @@ function handleRequest(mockAdapter, resolve, reject, config) {
   }
 }
 
+function getEffectiveDelay(adapter, handler){
+  var delayPerRequest = 0;
+  if(handler.length === 8){
+    delayPerRequest = handler[7];
+  }
+  return adapter.delayResponse + delayPerRequest
```
Comment (ctimmerm) [inline src/handle_request.js:145]: Instead of adding it to the global delay, I think it would be more intuitive if it overrode the global delay instead.

Comment (prashanth-92) [reply src/handle_request.js]: That actually makes sense, I have adjusted.

**Comment by prashanth-92** [comment]
Please re-review @ctimmerm

**Comment by prashanth-92** [comment]
Please re review.

**Comment by prashanth-92** [comment]
@ctimmerm  please review.

**Comment by prashanth-92** [comment]
@ctimmerm Can you please re review?

File: src/index.js [hunk 2]
```diff
@@ -89,6 +89,18 @@ VERBS.concat("any").forEach(function (method) {
       return _this;
     }
 
+    function replyWithDelay(delay, code, response, headers) {
+      var handler = [matcher, body, requestHeaders, code, response, headers, false, delay];
+      addHandler(method, _this.handlers, handler);
+      return _this;
+    }
+
+    function withDelayInMs(delay){
```
Comment (katreniak) [inline src/index.js:98]: Naming inconstency: above method `replyWithDelay` does not have `InMs` suffix.

Rename to `withDelay`?

**Comment by jshall** [comment]
I would also be interested to have `withDelay()` accept a function.  This would allow me to create a function that generates a random delay in a specific range.

**Comment by prashanth-92** [comment]
> I would also be interested to have `withDelay()` accept a function.  This would allow me to create a function that generates a random delay in a specific range.

The idea behind this pull request is to give a helper for introducing a delay time alone. Your new feature request can be addressed in another pull request. Feel free to open one.

**Comment by phiter** [comment]
Hey guys, is this getting merged or nah? I just found a use case where I need a delay for a specific request.

**Comment by prashanth-92** [comment]
@ctimmerm If there are no concerns, can we merge this PR? I think it will be useful.

**Comment by marcbachmann** [comment]
hmm, a force push merged this accidentally and can't reopen. I'll merge this manually

**Comment by marcbachmann** [comment]
hah. reopen button shows up again

**Comment by marcbachmann** [comment]
Released as v1.22.0

**Comment by jfstephe** [comment]
[marcbachmann](https://github.com/marcbachmann), [prashanth-92](https://github.com/prashanth-92) - I'm a bit late to the party but there's no documentation and I think the typescript definitions are missing...

**Comment by jfstephe** [comment]
Also IMHO the API for this is now unintuitive (it looked like @ctimmerm had a good suggestion). Why the change to make this return the reply function rather than the request handler?

**Comment by Glideh** [comment]
Interested about that too.
Need a doc & types up to date plz.

**Comment by seriouslag** [comment]
> Interested about that too. Need a doc & types up to date plz.

I opened a PR to add types to replyWithDelay
https://github.com/ctimmerm/axios-mock-adapter/pull/383

**Pull request closed without merging**